sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Throttle.java

```diff
@@ -0,0 +1,469 @@
+ * DoFn</a>. Next the transform applies {@link GlobalWindows} to the {@code PCollection<KV<Integer,
+ * List<RequestT>>>} prior to applying to the splittable DoFn. This splittable DoFn performs the
+ * actual work of throttling by holding the watermark and performing a {@link
+ * DoFn.ProcessContinuation#withResumeDelay(Duration)} of {@link Rate#getInterval()} if there are
+ * remaining elements to process. Finally, the transform applies the original input's {@link
+ * WindowingStrategy} to the returning {@code PCollection<RequestT>}.
+ */
+public class Throttle<RequestT> extends PTransform<PCollection<RequestT>, PCollection<RequestT>> {
+  static final String INPUT_ELEMENTS_COUNTER_NAME = "input_elements_count";
+  static final String OUTPUT_ELEMENTS_COUNTER_NAME = "output_elements_count";
+
+  /**
+   * Instantiates a {@link Throttle} with the maximumRate of {@link Rate} and without collecting
+   * metrics.
+   */
+  static <RequestT> Throttle<RequestT> of(Rate maximumRate) {
+    return new Throttle<>(Configuration.builder().setMaximumRate(maximumRate).build());
+  }
+
+  /** Returns {@link Throttle} with metrics collection turned on. */
+  Throttle<RequestT> withMetricsCollected() {
+    return new Throttle<>(configuration.toBuilder().setCollectMetrics(true).build());
+  }
+
+  /**
+   * Configures {@link Throttle} with additional parameters for use in streaming contexts. Calls
+   * {@link #withStreamingConfiguration(long, Duration)} with {@code null} {@link Duration}
+   * argument. See {@link #withStreamingConfiguration(long, Duration)} for more details.
+   */
+  Throttle<RequestT> withStreamingConfiguration(long bufferingSize) {
+    return withStreamingConfiguration(bufferingSize, null);
+  }
+
+  /**
+   * Configures {@link Throttle} for use with {@link PCollection.IsBounded#UNBOUNDED} {@link
+   * PCollection}s. In a streaming context, {@link Throttle} applies the input {@code
+   * PCollection<RequestT>} to {@link GroupIntoBatches} prior to throttling. Therefore, it requires
+   * additional configuration for how to handle streaming contexts.
+   */
+  Throttle<RequestT> withStreamingConfiguration(
+      long bufferingSize, @Nullable Duration maxBufferingDuration) {
+    return new Throttle<>(
+        configuration
+            .toBuilder()
+            .setStreamBufferingSize(bufferingSize)
+            .setStreamMaxBufferingDuration(maxBufferingDuration)
+            .build());
+  }
+
+  private final Configuration configuration;
+
+  private Throttle(Configuration configuration) {
+    this.configuration = configuration;
+  }
+
+  @Override
+  public PCollection<RequestT> expand(PCollection<RequestT> input) {
+    ListCoder<RequestT> listCoder = ListCoder.of(input.getCoder());
+    Coder<KV<Integer, List<RequestT>>> kvCoder = KvCoder.of(VarIntCoder.of(), listCoder);
+
+    PTransform<PCollection<KV<Integer, RequestT>>, PCollection<KV<Integer, Iterable<RequestT>>>>
+        groupingTransform = GroupByKey.create();
```

Comment:
Is there a reason we can't just always use GroupIntoBatches? I think it will tend to perform better because it won't require the whole previous stage to complete before we start sending requests

Reply:
Tests do not work in batch mode as it relies on event timers.

Reply:
Ok - I think this is a major performance problem since it will require all reads to complete before starting the throttling.